File: app/src/main/java/com/example/homestay/ui/admin/AdminBookingsActivity.java
package com.example.homestay.ui.admin;

import android.os.Bundle;
import android.text.*;
import android.view.View;
import android.widget.*;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.*;
import com.example.homestay.HomestayApplication;
import com.example.homestay.R;
import com.example.homestay.data.entity.*;
import com.example.homestay.data.model.*;
import com.example.homestay.data.repository.HomestayRepository;
import com.example.homestay.domain.BookingStatusPolicy;
import com.example.homestay.domain.CancellationPolicy;
import com.example.homestay.utils.*;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import java.text.*;
import java.util.*;

public class AdminBookingsActivity extends AppCompatActivity {
  public static final String EXTRA_BOOKING_CODE = "booking_code";

  private HomestayRepository repository;
  private AdminBookingAdapter adapter;
  private ProgressBar progress;
  private final List<AdminBookingData> bookings = new ArrayList<>();
  private String selectedStatus;

  @Override
  protected void onCreate(Bundle s) {
    super.onCreate(s);
    setContentView(R.layout.activity_admin_bookings);
    SystemBarUtils.keepContentBelowStatusBar(this);
    repository = ((HomestayApplication) getApplication()).getRepository();
    Toolbar toolbar = findViewById(R.id.toolbar);
    setSupportActionBar(toolbar);
    if (getSupportActionBar() != null) {
      getSupportActionBar().setDisplayHomeAsUpEnabled(false);
      getSupportActionBar().setTitle("Quản lý đặt chỗ");
    }
    setup();
    String bookingCode = getIntent().getStringExtra(EXTRA_BOOKING_CODE);
    if (bookingCode != null)
      ((TextInputEditText) findViewById(R.id.et_admin_search)).setText(bookingCode);
    AdminNavigationUtils.setup(this, R.id.admin_nav_bookings);
    load();
  }

  private void setup() {
    progress = findViewById(R.id.progress_bar);
    adapter =
        new AdminBookingAdapter(
            new ArrayList<>(), this::showStatus, this::showDelete, this::showDetails);
    RecyclerView list = findViewById(R.id.rv_bookings);
    list.setLayoutManager(new LinearLayoutManager(this));
    list.setAdapter(adapter);
    TextInputEditText search = findViewById(R.id.et_admin_search);
    search.addTextChangedListener(
        new TextWatcher() {
          public void beforeTextChanged(CharSequence s, int a, int b, int c) {}

          public void onTextChanged(CharSequence s, int a, int b, int c) {
            filter();
          }

          public void afterTextChanged(Editable e) {}
        });
    ((ChipGroup) findViewById(R.id.chip_booking_status))
        .setOnCheckedStateChangeListener(
            (g, ids) -> {
              int id = ids.isEmpty() ? View.NO_ID : ids.get(0);
              selectedStatus =
                  id == R.id.chip_pending
                      ? "pending"
                      : id == R.id.chip_confirmed
                          ? "confirmed"
                          : id == R.id.chip_checked_in
                              ? "checked_in"
                          : id == R.id.chip_completed
                              ? "completed"
                              : id == R.id.chip_cancelled ? "cancelled" : null;
              filter();
            });
  }

  private void load() {
    progress.setVisibility(View.VISIBLE);
    AppExecutors.io()
        .execute(
            () -> {
              try {
                Map<Long, Room> rooms = new HashMap<>();
                for (Room r : repository.getAllRoomsNow()) rooms.put(r.getId(), r);
                Map<Long, User> users = new HashMap<>();
                for (User u : repository.getAllUsersNow()) users.put(u.getId(), u);
                List<AdminBookingData> result = new ArrayList<>();
                for (Booking b : repository.getAllBookingsNow()) {
                  User u = users.get(b.getUserId());
                  Room r = rooms.get(b.getRoomId());
                  result.add(
                      new AdminBookingData(
                          String.valueOf(b.getId()),
                          u == null
                              ? null
                              : new AdminBookingUser(
                                  String.valueOf(u.getId()),
                                  u.getEmail(),
                                  u.getFullName(),
                                  u.getPhone()),
                          r == null
                              ? null
                              : new AdminBookingRoom(
                                  String.valueOf(r.getId()), r.getName(), r.getPrice()),
                          b.getCheckInDate(),
                          b.getCheckOutDate(),
                          b.getGuestCount(),
                          b.getTotalPrice(),
                          b.getStatus(),
                          b.getPaymentMethod(),
                          b.getPaymentStatus(),
                          b.getCreatedAt(),
                          b.getSlotId() == null ? null : String.valueOf(b.getSlotId())));
                }
                runOnUiThread(
                    () -> {
                      bookings.clear();
                      bookings.addAll(result);
                      filter();
                      progress.setVisibility(View.GONE);
                    });
              } catch (Exception e) {
                runOnUiThread(
                    () -> {
                      progress.setVisibility(View.GONE);
                      toast("Lỗi: " + e.getMessage());
                    });
              }
            });
  }

  private void filter() {
    if (adapter == null) return;
    String query =
        ((TextInputEditText) findViewById(R.id.et_admin_search)).getText() == null
            ? ""
            : ((TextInputEditText) findViewById(R.id.et_admin_search))
                .getText()
                .toString()
                .trim()
                .toLowerCase(Locale.ROOT);
    List<AdminBookingData> out = new ArrayList<>();
    for (AdminBookingData b : bookings) {
      String value = DisplayFormatter.bookingCode(b.getId(), b.getCreatedAt());
      String user =
          b.getUser() == null ? "" : b.getUser().getFullName() + " " + b.getUser().getEmail();
      String room = b.getRoom() == null ? "" : b.getRoom().getName();
      if ((query.isEmpty()
              || (user + " " + room + " " + value)
                  .toLowerCase(Locale.ROOT)
                  .contains(query.replace("#", "")))
          && (selectedStatus == null || selectedStatus.equals(b.getStatus()))) out.add(b);
    }
    adapter.updateBookings(out);
    findViewById(R.id.tv_empty).setVisibility(out.isEmpty() ? View.VISIBLE : View.GONE);
  }

  private void showStatus(AdminBookingData b) {
    List<String> values = new ArrayList<>(), labels = new ArrayList<>();
    if ("cancelled".equalsIgnoreCase(b.getStatus())
        && "REFUND_PENDING".equals(b.getPaymentStatus())) {
      new MaterialAlertDialogBuilder(this)
          .setTitle("Xác nhận hoàn tiền")
          .setMessage("Đánh dấu khoản hoàn của booking này đã được xử lý?")
          .setNegativeButton("Để sau", null)
          .setPositiveButton("Đã hoàn tiền", (dialog, which) -> markRefundProcessed(b.getId()))
          .show();
      return;
    }
    if ("pending".equalsIgnoreCase(b.getStatus())) {
      values.add("confirmed");
      labels.add("Đã xác nhận");
      values.add("cancelled");
      labels.add("Đã hủy");
    } else if ("confirmed".equalsIgnoreCase(b.getStatus())) {
      if (BookingStatusPolicy.canComplete(
          b.getStatus(), b.getCheckOutDate(), System.currentTimeMillis())) {
        values.add("completed");
        labels.add("Hoàn thành");
      }
      values.add("cancelled");
      labels.add("Đã hủy");
    }
    if (values.isEmpty()) {
      toast("Booking đã ở trạng thái cuối");
      return;
    }
    new MaterialAlertDialogBuilder(this)
        .setTitle("Đổi trạng thái booking")
        .setSingleChoiceItems(
            labels.toArray(new String[0]),
            -1,
            (d, w) -> {
              if ("cancelled".equals(values.get(w))) showAdminCancellation(b);
              else updateStatus(b.getId(), values.get(w));
              d.dismiss();
            })
        .setNegativeButton("Hủy", null)
        .show();
  }

  private void markRefundProcessed(String id) {
    AppExecutors.io()
        .execute(
            () -> {
              Booking booking = repository.getBookingById(Long.parseLong(id));
              if (booking != null) {
                String paymentStatus =
                    booking.getRefundAmount() < booking.getTotalPrice()
                        ? "PARTIALLY_REFUNDED"
                        : "REFUNDED";
                repository.updateBooking(booking.withPaymentStatus(paymentStatus));
              }
              runOnUiThread(
                  () -> {
                    toast("Đã cập nhật trạng thái hoàn tiền");
                    load();
                  });
            });
  }

  private void showAdminCancellation(AdminBookingData data) {
    TextInputEditText reason = new TextInputEditText(this);
    reason.setHint("Lý do hủy từ quản trị viên");
    reason.setText("Phòng không thể tiếp nhận khách");
    int padding = (int) (24 * getResources().getDisplayMetrics().density);
    android.widget.FrameLayout wrapper = new android.widget.FrameLayout(this);
    wrapper.setPadding(padding, 0, padding, 0);
    wrapper.addView(reason);
    new MaterialAlertDialogBuilder(this)
        .setTitle("Hủy booking")
        .setMessage("Khách sẽ nhận được lý do hủy; chỉ booking đã thanh toán mới phát sinh hoàn tiền.")
        .setView(wrapper)
        .setNegativeButton("Quay lại", null)
        .setPositiveButton(
            "Xác nhận hủy",
            (dialog, which) -> {
              String value =
                  reason.getText() == null ? "Hủy bởi quản trị viên" : reason.getText().toString().trim();
              if (value.isEmpty()) value = "Hủy bởi quản trị viên";
              cancelByAdmin(data.getId(), value);
            })
        .show();
  }

  private void cancelByAdmin(String id, String reason) {
    AppExecutors.io()
        .execute(
            () -> {
              try {
                Booking booking = repository.getBookingById(Long.parseLong(id));
                if (booking != null) {
                  double refund =
                      "PAID".equals(booking.getPaymentStatus())
                          ? CancellationPolicy.refund(
                              booking.getTotalPrice(),
                              booking.getCheckInDate(),
                              System.currentTimeMillis())
                          : 0;
                  repository.updateBooking(
                      booking.cancelled(
                          "Quản trị viên: " + reason, refund, System.currentTimeMillis()));
                }
                runOnUiThread(
                    () -> {
                      toast("Đã hủy booking và gửi thông báo cho khách");
                      load();
                    });
              } catch (Exception error) {
                runOnUiThread(() -> toast("Không thể hủy: " + error.getMessage()));
              }
            });
  }

  private void updateStatus(String id, String status) {
    AppExecutors.io()
        .execute(
            () -> {
              try {
                Booking b = repository.getBookingById(Long.parseLong(id));
                if (b != null) {
                  Booking updated = b.withStatus(status, b.getPaymentMethod());
                  if ("completed".equals(status)) updated = updated.withPaymentStatus("PAID");
                  repository.updateBooking(updated);
                }
                runOnUiThread(
                    () -> {
                      toast("Cập nhật trạng thái thành công!");
                      load();
                    });
              } catch (Exception error) {
                runOnUiThread(() -> toast("Không thể cập nhật: " + error.getMessage()));
              }
            });
  }

  private void showDelete(AdminBookingData b) {
    new MaterialAlertDialogBuilder(this)
        .setTitle("Xóa booking")
        .setMessage("Bạn có chắc chắn muốn xóa booking này?")
        .setPositiveButton(
            "Xóa",
            (d, w) ->
                AppExecutors.io()
                    .execute(
                        () -> {
                          Booking value = repository.getBookingById(Long.parseLong(b.getId()));
                          boolean deleted = value != null && repository.deleteBooking(value);
                          runOnUiThread(
                              () -> {
                                if (deleted) {
                                  adapter.removeBooking(b.getId());
                                  toast("Xóa booking đã hủy thành công!");
                                } else toast("Chỉ có thể xóa booking đã hủy");
                              });
                        }))
        .setNegativeButton("Hủy", null)
        .show();
  }

  private void showDetails(AdminBookingData b) {
    SimpleDateFormat f = new SimpleDateFormat("dd/MM/yyyy HH:mm", new Locale("vi", "VN"));
    long nights =
        Math.max(1, com.example.homestay.domain.BookingCalculator.nights(b.getCheckInDate(), b.getCheckOutDate()));
    String room = b.getRoom() == null ? "N/A" : b.getRoom().getName(),
        user = b.getUser() == null ? "N/A" : b.getUser().getFullName(),
        email = b.getUser() == null ? "N/A" : b.getUser().getEmail(),
        phone = b.getUser() == null ? "N/A" : b.getUser().getPhone();
    new MaterialAlertDialogBuilder(this)
        .setTitle("Chi tiết " + DisplayFormatter.bookingCode(b.getId(), b.getCreatedAt()))
        .setMessage(
            "Phòng: "
                + room
                + "\nKhách: "
                + user
                + "\nEmail: "
                + email
                + "\nĐiện thoại: "
                + phone
                + "\n\nNhận phòng: "
                + f.format(new Date(b.getCheckInDate()))
                + "\nTrả phòng: "
                + f.format(new Date(b.getCheckOutDate()))
                + "\nThời gian lưu trú: "
                + nights
                + " đêm\nSố khách: "
                + b.getGuestCount()
                + "\n\nTổng tiền: "
                + NumberFormat.getNumberInstance(new Locale("vi", "VN"))
                    .format((long) b.getTotalPrice())
                + " đ\nTrạng thái: "
                + b.getStatus())
        .setPositiveButton("Đóng", null)
        .show();
  }

  private void toast(String s) {
    Toast.makeText(this, s, Toast.LENGTH_SHORT).show();
  }
}

File: app/src/main/java/com/example/homestay/AdminBookingsActivity.kt
package com.example.homestay

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.lifecycleScope
import androidx.core.widget.addTextChangedListener
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.chip.ChipGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.homestay.data.model.AdminBookingData
import com.example.homestay.data.model.AdminBookingRoom
import com.example.homestay.data.model.AdminBookingUser
import com.example.homestay.ui.admin.AdminBookingAdapter
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.launch
import kotlinx.coroutines.flow.first

class AdminBookingsActivity : AppCompatActivity() {
    private val repository by lazy { (application as HomestayApplication).repository }
    
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: AdminBookingAdapter
    private lateinit var progressBar: ProgressBar
    private var bookings = mutableListOf<AdminBookingData>()
    private var selectedStatus: String? = null
    private var hasPaymentMethod: Boolean? = null
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_admin_bookings)
        
        setupToolbar()
        setupViews()
        loadBookings()
    }
    
    private fun setupToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = "Quản lý Bookings"
    }
    
    private fun setupViews() {
        recyclerView = findViewById(R.id.rv_bookings)
        progressBar = findViewById(R.id.progress_bar)
        
        adapter = AdminBookingAdapter(
            bookings = bookings,
            onChangeStatusClick = { booking -> showChangeStatusDialog(booking) },
            onDeleteClick = { booking -> showDeleteConfirmDialog(booking) },
            onDetailsClick = { booking -> showBookingDetails(booking) }
        )
        
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter
        findViewById<TextInputEditText>(R.id.et_admin_search).addTextChangedListener { applyBookingFilters() }
        findViewById<ChipGroup>(R.id.chip_booking_status).setOnCheckedStateChangeListener { _, ids ->
            selectedStatus = when (ids.firstOrNull()) {
                R.id.chip_pending -> "pending"
                R.id.chip_confirmed -> "confirmed"
                R.id.chip_completed -> "completed"
                R.id.chip_cancelled -> "cancelled"
                else -> null
            }
            applyBookingFilters()
        }
        findViewById<ChipGroup>(R.id.chip_payment_status).setOnCheckedStateChangeListener { _, ids ->
            hasPaymentMethod = when (ids.firstOrNull()) {
                R.id.chip_payment_set -> true; R.id.chip_payment_unset -> false; else -> null
            }
            applyBookingFilters()
        }
    }
    
    private fun loadBookings() {
        progressBar.visibility = View.VISIBLE
        lifecycleScope.launch {
            try {
                val localBookings = repository.getAllBookings().first()
                val localRooms = repository.getAllRooms().first().associateBy { it.id }
                val localUsers = repository.getAllUsers().first().associateBy { it.id }
                val bookingsList = localBookings.map { booking ->
                    val user = localUsers[booking.userId]
                    val room = localRooms[booking.roomId]
                    AdminBookingData(
                        id = booking.id.toString(),
                        user = user?.let { AdminBookingUser(it.id.toString(), it.email, it.fullName, it.phone) },
                        room = room?.let { AdminBookingRoom(it.id.toString(), it.name, it.price) },
                        checkInDate = booking.checkInDate, checkOutDate = booking.checkOutDate,
                        guestCount = booking.guestCount, totalPrice = booking.totalPrice,
                        status = booking.status, paymentMethod = booking.paymentMethod,
                        createdAt = booking.createdAt, slotId = booking.slotId?.toString()
                    )
                }
                bookings.clear()
                bookings.addAll(bookingsList)
                applyBookingFilters()
            } catch (e: Exception) {
                android.util.Log.e("AdminBookings", "Error: ${e.message}", e)
                Toast.makeText(this@AdminBookingsActivity, "Lỗi: ${e.message}", Toast.LENGTH_SHORT).show()
            } finally {
                progressBar.visibility = View.GONE
            }
        }
    }

    private fun applyBookingFilters() {
        if (!::adapter.isInitialized) return
        val query = findViewById<TextInputEditText>(R.id.et_admin_search).text?.toString().orEmpty().trim()
        val filtered = bookings.filter { booking ->
            val matchesText = query.isBlank() || booking.user?.fullName.orEmpty().contains(query, true) ||
                booking.user?.email.orEmpty().contains(query, true) || booking.room?.name.orEmpty().contains(query, true)
            matchesText && (selectedStatus == null || booking.status == selectedStatus) &&
                (hasPaymentMethod == null || (booking.paymentMethod != null) == hasPaymentMethod)
        }
        adapter.updateBookings(filtered)
        findViewById<View>(R.id.tv_empty).visibility = if (filtered.isEmpty()) View.VISIBLE else View.GONE
    }
    
    private fun showChangeStatusDialog(booking: AdminBookingData) {
        val transitions = when (booking.status.lowercase()) {
            "pending" -> listOf("confirmed" to "Đã xác nhận", "cancelled" to "Đã hủy")
            "confirmed" -> listOf("completed" to "Hoàn thành", "cancelled" to "Đã hủy")
            else -> emptyList()
        }
        if (transitions.isEmpty()) {
            Toast.makeText(this, "Booking này đã ở trạng thái cuối", Toast.LENGTH_SHORT).show()
            return
        }
        val statuses = transitions.map { it.first }.toTypedArray()
        val statusLabels = transitions.map { it.second }.toTypedArray()
        
        MaterialAlertDialogBuilder(this)
            .setTitle("Đổi trạng thái booking")
            .setSingleChoiceItems(statusLabels, -1) { dialog, which ->
                val newStatus = statuses[which]
                if (newStatus != booking.status) {
                    updateBookingStatus(booking.id, newStatus)
                }
                dialog.dismiss()
            }
            .setNegativeButton("Hủy", null)
            .show()
    }

    private fun showBookingDetails(booking: AdminBookingData) {
        val formatDate = java.text.SimpleDateFormat("dd/MM/yyyy HH:mm", java.util.Locale("vi", "VN"))
        val nights = ((booking.checkOutDate - booking.checkInDate) / 86_400_000L).coerceAtLeast(1)
        val price = java.text.NumberFormat.getNumberInstance(java.util.Locale("vi", "VN")).format(booking.totalPrice.toLong())
        MaterialAlertDialogBuilder(this)
            .setTitle("Chi tiết booking #${booking.id}")
            .setMessage("Phòng: ${booking.room?.name ?: "N/A"}\nKhách: ${booking.user?.fullName ?: "N/A"}\nEmail: ${booking.user?.email ?: "N/A"}\nĐiện thoại: ${booking.user?.phone ?: "N/A"}\n\nNhận phòng: ${formatDate.format(java.util.Date(booking.checkInDate))}\nTrả phòng: ${formatDate.format(java.util.Date(booking.checkOutDate))}\nThời gian lưu trú: $nights đêm\nSố khách: ${booking.guestCount}\nSlot: ${booking.slotId ?: "Cả phòng"}\n\nTổng tiền: $price đ\nThanh toán: ${booking.paymentMethod ?: "Chưa chọn"}\nTrạng thái: ${booking.status}\nNgày tạo: ${formatDate.format(java.util.Date(booking.createdAt))}")
            .setPositiveButton("Đóng", null)
            .show()
    }
    
    private fun updateBookingStatus(bookingId: String, newStatus: String) {
        lifecycleScope.launch {
            try {
                val booking = repository.getBookingById(bookingId.toLong()) ?: return@launch
                repository.updateBooking(booking.copy(status = newStatus))
                Toast.makeText(this@AdminBookingsActivity, "Cập nhật trạng thái thành công!", Toast.LENGTH_SHORT).show()
                loadBookings()
            } catch (e: Exception) {
                Toast.makeText(this@AdminBookingsActivity, "Lỗi: ${e.message}", Toast.LENGTH_SHORT).show()
            }
        }
    }
    
    private fun showDeleteConfirmDialog(booking: AdminBookingData) {
        MaterialAlertDialogBuilder(this)
            .setTitle("Xóa booking")
            .setMessage("Bạn có chắc chắn muốn xóa booking này?")
            .setPositiveButton("Xóa") { _, _ ->
                deleteBooking(booking.id)
            }
            .setNegativeButton("Hủy", null)
            .show()
    }
    
    private fun deleteBooking(bookingId: String) {
        lifecycleScope.launch {
            try {
                val booking = repository.getBookingById(bookingId.toLong()) ?: return@launch
                repository.deleteBooking(booking)
                Toast.makeText(this@AdminBookingsActivity, "Xóa booking thành công!", Toast.LENGTH_SHORT).show()
                adapter.removeBooking(bookingId)
            } catch (e: Exception) {
                Toast.makeText(this@AdminBookingsActivity, "Lỗi: ${e.message}", Toast.LENGTH_SHORT).show()
            }
        }
    }
    
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}
